src/databricks/labs/ucx/progress/install.py

```diff
@@ -0,0 +1,75 @@
+import datetime as dt
+import logging
+from dataclasses import dataclass
+
+from databricks.labs.lsql.backends import Dataclass, SqlBackend
+from databricks.sdk.errors import InternalError
+from databricks.sdk.retries import retried
+
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class HistoricalRecord:
+    workspace_id: int
+    """The identifier of the workspace where this record was generated."""
+
+    run_id: int
+    """An identifier of the workflow run that generated this record."""
+
+    snapshot_id: int
+    """An identifier that is unique to the records produced for a given snapshot."""
+
+    run_start_time: dt.datetime
+    """When this record was generated."""
+
+    object_type: str
+    """The inventory table for which this record was generated."""
+
+    object_type_version: int
+    """Versioning of inventory table, for forward compatibility."""
+
+    object_id: list[str]
```

Comment:
We have a few things where the object_id is a composite from a few fields. My initial thinking here is that just using them as the `object_id` side-steps having to deal with corner cases like "what happens when I join these fields to make a string but the fields can contain the separator?".

I'm going through the various crawler record types now to see what this looks like in practice.

Reply:
please be concrete about what things do require this.

Reply:
We have the following classes that use composite keys:
 * `Grant`: `object_type`, `object_key`, `action_type`, `principal`
 * `Permissions`: `object_type`, `object_id`
 * `ReconResult`: `src_schema`, `src_table`
 * `Table`: `catalog`, `database`, `name`
 * `TableMigrationStatus`: `src_schema`, `src_table`
 * `TableSize`: `catalog`, `database`, `name`
 * `Udf`: `catalog`, `database`, `name`

(There's also the special case of `SubmitRunInfo` whose key, `run_ids`, is already a pseudo-composite key because it contains a JSON-encoded array of run identifiers.)

If we are going to just use `str` then I propose that the encoding is just the JSON-encoding of these fields as an array.

An alternative would be to use `map[str,str]`.

If we want to stick with `str` but not use JSON-encoding then we will need to come up with our composite encoding scheme, bearing in mind that the field values aren't under our control and an escaping mechanism will be needed.

Reply:
Following discussion, #2761 has been created. The broad outcome of that discussion was that:
 - Every object needs to have an identifiable owner. This is mandatory.
 - Often the owner is indeed a composite of other attributes.
 - Encoding/decoding the owner will be type-specific.
 - Therefore `str` is fine.

Reply:
@nfx: What do you think?

Reply:
okay, ~7 cases our of ~21 crawlers (33%) is a good share to have `object_id: list[str]`.